packages/plugin-node-modules/sources/NodeModulesLinker.ts

```diff
@@ -1,260 +1,147 @@
-        if (options && options.ignoreStateFileReadErrors) {
-          // Ignore errors if state file is absent
-          this.cachedLocatorMap = new Map();
-        } else {
-          throw e;
-        }
-      }
-      this.cachedLocationMap = new Map();
-      for (const [locatorKey, val] of this.cachedLocatorMap) {
-        const locator = structUtils.tryParseLocator(locatorKey)!;
-        for (const location of val.locations) {
-          this.cachedLocationMap.set(ppath.join(rootPath, location), locator);
-        }
-      }
-    }
-
-    return {locatorMap: this.cachedLocatorMap, locationMap: this.cachedLocationMap!};
-  }
-
   supportsPackage(pkg: Package, opts: MinimalLinkOptions) {
-    return opts.project.configuration.get('nodeLinker') === 'node-modules';
+    return opts.project.configuration.get(`nodeLinker`) === `node-modules`;
   }
 
   async findPackageLocation(locator: Locator, opts: LinkOptions) {
-    const rootPath = opts.project.cwd;
-    const {locatorMap} = await this.getLocatorMap(rootPath);
-    const locatorInfo = locatorMap.get(structUtils.stringifyLocator(locator));
+    const installState = await findInstallState(opts.project, {unrollAliases: true});
+    if (installState === null)
+      throw new UsageError(`Couldn't find the node_modules state file - running an install might help (findPackageLocation)`);
 
+    const locatorInfo = installState.get(structUtils.stringifyLocator(locator));
     if (!locatorInfo)
       throw new UsageError(`Couldn't find ${structUtils.prettyLocator(opts.project.configuration, locator)} in the currently installed node_modules map - running an install might help`);
 
-    return ppath.join(rootPath, locatorInfo.locations[0]);
+    return locatorInfo.locations[0];
   }
 
   async findPackageLocator(location: PortablePath, opts: LinkOptions) {
-    const rootPath = opts.project.cwd;
-    const {locationMap} = await this.getLocatorMap(rootPath, {ignoreStateFileReadErrors: true});
-    return locationMap.get(location) || null;
-  }
-
-  makeInstaller(opts: LinkOptions) {
-    return new NodeModulesInstaller(opts, () => this.getLocatorMap(opts.project.cwd, {reread: true}));
-  }
-}
-
-class NodeModulesInstaller implements Installer {
-  private readonly packageRegistry: PackageRegistry = new Map();
+    const installState = await findInstallState(opts.project, {unrollAliases: true});
+    if (installState === null)
+      return null;
 
-  private readonly unpluggedPaths: Set<string> = new Set();
-  private readonly blacklistedPaths: Set<string> = new Set();
-  private readonly installedPackages: Map<LocatorKey, PortablePath> = new Map();
-  private readonly onFinalizeInstall: () => Promise<any>;
+    const locationMap = getLocationMap(installState);
 
-  private readonly opts: LinkOptions;
+    // TODO: Doesn't support nested paths; we'd need to do something similar
+    // to the `findPackageLocator` in the PnP API.
+    //
+    // ex: /.../node_modules/foo/subdirectory/another/file.js
```

Comment:
I'm not sure I understand this. What about just doing `ppath.resolve` on this path and checking absolute path in `locationMap`?

Reply:
Assuming that `location` is `/root/node_modules/foo/subdirectory/another/` (which would happen if a package was running `yarn run` within one of its subdirectories), it won't be in the location map (but `/root/node_modules/foo` would).

So either we need to find the first subpath that is found in the map, or we can make that slightly faster by only checking the first `node_modules` ancestor. But in the current state, it won't work with subdirectories.

Reply:
I think we just need to find the last `node_modules` segment in the path, then check if the next segment is scope or not and depending on that check either `/.../node_modules/@foo/bar` or `/../node_modules/foo` in the location map and then add the rest of the original request to that location